### DISCUSSION TITLE
Thermal-Hydraulics model problem: using the iteration algorithm to calculate Tw

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#25447] [Q&A Modules: General] [answered]
Hi all,

I would like to use the Wolf-McCarthy correlation to calculate Tw by giving a constant wall heat flux in ThermalHydraulic model. The way is using this formula: Tw = qw / Hw + Tf.

But the Wolf-McCarthy correlation needs Tw to calculate Hw. So both Tw and Hw are unknown. Therefore, I need to use the iteration algorithm to calculate this.

I changed the function in ADWallHeatTransferCoefficientWolfMcCarthyMaterial.C like this:

```
void
ADWallHeatTransferCoefficientWolfMcCarthyMaterial::computeQpProperties()
{
  ADReal qw = 1.0e5; // given value
  ADReal deta_Tw = 10.0; // just give an initial value
  ADReal Tw_init = 500.0; // just give an initial value
  ADReal Tw_itera;
  ADReal Tw_calc = Tw_init;
  ADReal Pr;
  ADReal Re;
  ADReal Nu;
  while(deta_Tw > 0.01)
  {
    
    Tw_itera = Tw_calc;
    Pr = THM::Prandtl(_cp[_qp], _mu[_qp], _k[_qp]);
    Re = std::max(1.0, THM::Reynolds(1., _rho[_qp], _vel[_qp], _D_h[_qp], _mu[_qp]));
    Nu = 0.025 * std::pow(Re, 0.8) * std::pow(Pr, 0.4) *
                std::pow(std::max(_T[_qp], Tw_itera) / _T[_qp], -0.55);
    _Hw[_qp] = THM::wallHeatTransferCoefficient(Nu, _k[_qp], _D_h[_qp]);
    
    Tw_calc = qw / _Hw[_qp] + _T[_qp];

    deta_Tw = std::abs(Tw_itera - Tw_calc);
  }

  _T_wall[_qp] = Tw_calc;
}
```

But when I run the case, It is always in the state of "Performing Initial Setup...".

![image](https://github.com/idaholab/moose/assets/118275988/07174bc4-a846-446c-9069-4342939c5e6c)

What is the problem? Could anyone kindly tell me?

Best regards!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Your algorithm is not converging and you're stuck in an infinite loop would be my guess.
You should:
- print the values or use a debugger to see what is going on in the loop
- consider using Newton's method instead of Picard as it converges faster

Guillaume

**Reply by SomnusYu** [reply]
Thank you for your suggestions.

I have fixed this problem. The problem is that I give the wrong boundary condition.

By the way, what Newton's method should be in this example?

Thank you very much!

**Reply by GiudGiud** [reply]
Newton-Raphson ?
The basic one for solving f(x) = 0

**Reply by SomnusYu** [reply]
I will have a try. Thank you very much!
